tidb-latency-formula.md

```diff
@@ -0,0 +1,531 @@
+```
+
+After [getting snapshot](#tikv-snapshot), TiKV read multi values from the same snapshot. The read duration is same with point get.
+When loading data from disk, the average duration can be calculated by `tikv_storage_rocksdb_perf` with `req="batch_get"`.
+
+### Table Scan & Index Scan
+
+```
+tidb_session_execute_duration_seconds{type="general"} =
+    pd_client_cmd_handle_cmds_duration_seconds{type="wait"} +
+    n * (
+        tidb_distsql_handle_query_duration_seconds{sql_type="general"}
+    )
+    tidb_distsql_handle_query_duration_seconds{sql_type="general"} <= send request duration
+```
+
+Table scan and index scan are processed in the same way. `n` is the distributed task count.
+Because coprocessor execution and data responsing to client are in different threads, `tidb_distsql_handle_query_duration_seconds{sql_type="general"}` is the wait time and it's less than the send request duration.
+
+```
+send request duration =
+    tidb_tikvclient_batch_wait_duration +
+    tidb_tikvclient_batch_send_latency +
+    tikv_grpc_msg_duration_seconds{type="coprocessor"} +
+    tidb_tikvclient_rpc_net_latency_seconds{store="?"}
+
+tikv_grpc_msg_duration_seconds{type="coprocessor"} =
+    tikv_coprocessor_request_wait_seconds{type="snapshot"} +
+    tikv_coprocessor_request_wait_seconds{type="schedule"} +
+    tikv_coprocessor_request_handler_build_seconds{type="index/select"} +
+    tikv_coprocessor_request_handle_seconds{type="index/select"}
+```
+
+In TiKV, table scan is the type `select` and index scan is the type `index`, the duration details are same.
+
+### Index Look Up
+
+```
+tidb_session_execute_duration_seconds{type="general"} =
+    pd_client_cmd_handle_cmds_duration_seconds{type="wait"} +
+    n * (
+        tidb_distsql_handle_query_duration_seconds{sql_type="general"}
+    ) +
+    m * (
+        tidb_distsql_handle_query_duration_seconds{sql_type="general"}
+    )
+```
+
+Index look up combines index scan and table scan, they are processed in a pipelined way.
+
+## Write Queries
+
+Write queries are much more complex, there are some variants.
+
+|| Pessimistic Txn | Optimistic Txn |
+|-|-|-|
+| Auto-commit | execute + lock + commit | execute + commit |
+| Non auto-commit | execute + lock | execute |
+
+We divide the write into 3 phases:
+
+- execute phase, execute and write mutation into the memory of TiDB.
+- lock phase, acquire pessimistic locks for the execution result.
+- commit phase, commit the transaction via 2PC protocol.
+
+In execution phase, TiDB manipulate data in memory, the main latency comes from reading the required data.
+Like update and delete, TiDB read data from TiKV first, then update or delete the row in-mem.
+
+The exception is cursor operations(`select for update`), cursor operation with point get and batch point get, which performs read and lock in a single RPC.
```

Comment:
Why is `select for update` called cursor operations?

Reply:
Well, I thought they were the same before(they behave the same way), I will change it.